File: recipes/claude/06-handle-stop-reasons/python/stop_reasons.py
"""
Stop Reasons and Max Token Truncation Handler for Anthropic Claude.

This module provides production patterns for inspecting Claude stop reasons,
handling cutoff boundaries, and orchestrating continuation loops when
stop_reason == "max_tokens".
"""

from __future__ import annotations

import os
import sys
from typing import Generator, List, Optional
import anthropic
from anthropic.types import Message, TextBlock


def inspect_stop_reason(message: Message) -> None:
    """Log and validate the stop_reason on a completed Messages API response."""
    print(f"Message ID: {message.id}")
    print(f"Model: {message.model}")
    print(f"Stop Reason: {message.stop_reason}")
    print(f"Stop Sequence: {message.stop_sequence}")
    print(f"Usage: Input={message.usage.input_tokens}, Output={message.usage.output_tokens}")

    if message.stop_reason == "end_turn":
        print("Status: Natural conclusion reached.")
    elif message.stop_reason == "max_tokens":
        print("Status: Output truncated due to max_tokens limit.")
    elif message.stop_reason == "stop_sequence":
        print(f"Status: Execution stopped by custom stop sequence: {message.stop_sequence}")
    elif message.stop_reason == "tool_use":
        print("Status: Model stopped to invoke a tool.")
    else:
        print(f"Status: Unexpected stop reason: {message.stop_reason}")


def generate_with_continuation(
    client: anthropic.Anthropic,
    prompt: str,
    system: Optional[str] = None,
    model: str = "claude-3-5-sonnet-20241022",
    max_tokens_per_call: int = 1000,
    max_continuation_rounds: int = 5,
) -> str:
    """
    Execute a prompt and automatically loop if truncated by max_tokens.

    When stop_reason is "max_tokens", the accumulated response text is appended
    as an assistant turn, followed by a user prompt requesting continuation
    from the exact cutoff point without repeating previous text.
    """
    messages: List[dict] = [{"role": "user", "content": prompt}]
    accumulated_segments: List[str] = []
    round_count = 0

    while round_count < max_continuation_rounds:
        round_count += 1
        kwargs = {
            "model": model,
            "max_tokens": max_tokens_per_call,
            "messages": messages,
        }
        if system:
            kwargs["system"] = system

        response: Message = client.messages.create(**kwargs)

        # Extract text content from the blocks
        round_text = "".join(
            block.text for block in response.content if isinstance(block, TextBlock)
        )
        accumulated_segments.append(round_text)

        print(
            f"[Round {round_count}] Tokens: in={response.usage.input_tokens}, "
            f"out={response.usage.output_tokens} | Stop Reason: {response.stop_reason}"
        )

        if response.stop_reason == "end_turn":
            # Natural completion
            break
        elif response.stop_reason == "max_tokens":
            # Output truncated, prepare next turn
            messages.append({"role": "assistant", "content": round_text})
            messages.append(
                {
                    "role": "user",
                    "content": (
                        "Continue generating your output from the exact point where you "
                        "were cut off. Do not repeat any prior text, acknowledge this instruction, "
                        "or prepend commentary. Resume directly."
                    ),
                }
            )
        elif response.stop_reason == "stop_sequence":
            print(f"Terminated by custom sequence: {response.stop_sequence}")
            break
        elif response.stop_reason == "tool_use":
            print("Encountered tool invocation during text continuation loop.")
            break
        else:
            print(f"Unknown stop reason: {response.stop_reason}. Exiting continuation.")
            break

    return "".join(accumulated_segments)


def stream_with_stop_reason_check(
    client: anthropic.Anthropic,
    prompt: str,
    max_tokens: int = 500,
    model: str = "claude-3-5-sonnet-20241022",
) -> str:
    """Stream response tokens in real-time and extract final stop reason cleanly."""
    collected_text = []

    with client.messages.stream(
        model=model,
        max_tokens=max_tokens,
        messages=[{"role": "user", "content": prompt}],
    ) as stream:
        for text in stream.text_stream:
            print(text, end="", flush=True)
            collected_text.append(text)

        final_message = stream.get_final_message()
        print(f"\n[Stream Completed] Stop Reason: {final_message.stop_reason}")
        print(f"[Tokens] Input={final_message.usage.input_tokens}, Output={final_message.usage.output_tokens}")

    return "".join(collected_text)


def main() -> None:
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        print("ERROR: ANTHROPIC_API_KEY is not set in environment.", file=sys.stderr)
        sys.exit(1)

    client = anthropic.Anthropic(api_key=api_key)

    print("=== Demo 1: Triggering and Inspecting max_tokens Cutoff ===")
    short_resp = client.messages.create(
        model="claude-3-5-sonnet-20241022",
        max_tokens=20,
        messages=[
            {
                "role": "user",
                "content": "Explain the raft consensus protocol step by step.",
            }
        ],
    )
    inspect_stop_reason(short_resp)

    print("\n=== Demo 2: Triggering stop_sequence ===")
    seq_resp = client.messages.create(
        model="claude-3-5-sonnet-20241022",
        max_tokens=100,
        stop_sequences=["--END--"],
        messages=[
            {
                "role": "user",
                "content": "Count from 1 to 10. Place --END-- after number 4.",
            }
        ],
    )
    inspect_stop_reason(seq_resp)

    print("\n=== Demo 3: Automated Continuation Loop on max_tokens ===")
    continuation_result = generate_with_continuation(
        client=client,
        prompt="Write a detailed, structured 4-section guide on database indexing strategies.",
        max_tokens_per_call=80,
        max_continuation_rounds=3,
    )
    print("\n--- Final Consolidated Text (Preview first 200 chars) ---")
    print(continuation_result[:200] + "...")


if __name__ == "__main__":
    main()

File: recipes/claude/06-handle-stop-reasons/typescript/stop_reasons.ts
import Anthropic from '@anthropic-ai/sdk';
import type { Message, TextBlock } from '@anthropic-ai/sdk/resources/messages';

interface ContinuationOptions {
  client: Anthropic;
  prompt: string;
  system?: string;
  model?: string;
  maxTokensPerCall?: number;
  maxRounds?: number;
}

/**
 * Inspect and log stop_reason metadata from an Anthropic Message response.
 */
export function inspectStopReason(message: Message): void {
  console.log(`Message ID: ${message.id}`);
  console.log(`Model: ${message.model}`);
  console.log(`Stop Reason: ${message.stop_reason}`);
  console.log(`Stop Sequence: ${message.stop_sequence ?? 'none'}`);
  console.log(`Usage: Input=${message.usage.input_tokens}, Output=${message.usage.output_tokens}`);

  switch (message.stop_reason) {
    case 'end_turn':
      console.log('Status: Model finished natural completion.');
      break;
    case 'max_tokens':
      console.log('Status: Generation cut off due to max_tokens limit.');
      break;
    case 'stop_sequence':
      console.log(`Status: Custom stop sequence matched: ${message.stop_sequence}`);
      break;
    case 'tool_use':
      console.log('Status: Execution halted for tool invocation.');
      break;
    default:
      console.log(`Status: Unrecognized stop reason: ${message.stop_reason}`);
  }
}

/**
 * Execute a completion prompt and automatically continue across sequential
 * turns if the model terminates with stop_reason === 'max_tokens'.
 */
export async function generateWithContinuation({
  client,
  prompt,
  system,
  model = 'claude-3-5-sonnet-20241022',
  maxTokensPerCall = 1000,
  maxRounds = 5,
}: ContinuationOptions): Promise<string> {
  const messages: Array<{ role: 'user' | 'assistant'; content: string }> = [
    { role: 'user', content: prompt },
  ];
  const accumulatedSegments: string[] = [];
  let round = 0;

  while (round < maxRounds) {
    round += 1;

    const response = await client.messages.create({
      model,
      max_tokens: maxTokensPerCall,
      messages,
      ...(system ? { system } : {}),
    });

    const roundText = response.content
      .filter((block): block is TextBlock => block.type === 'text')
      .map((block) => block.text)
      .join('');

    accumulatedSegments.push(roundText);

    console.log(
      `[Round ${round}] Tokens: in=${response.usage.input_tokens}, out=${response.usage.output_tokens} | Stop Reason: ${response.stop_reason}`
    );

    if (response.stop_reason === 'end_turn') {
      break;
    } else if (response.stop_reason === 'max_tokens') {
      // Append assistant's partial output and prompt for seamless resumption
      messages.push({ role: 'assistant', content: roundText });
      messages.push({
        role: 'user',
        content:
          'Continue generating your output from the exact point where you were cut off. Do not repeat any prior text, acknowledge this instruction, or prepend commentary. Resume directly.',
      });
    } else if (response.stop_reason === 'stop_sequence') {
      console.log(`Halted on custom stop sequence: ${response.stop_sequence}`);
      break;
    } else if (response.stop_reason === 'tool_use') {
      console.log('Halted on tool invocation.');
      break;
    } else {
      console.log(`Halted on unhandled stop reason: ${response.stop_reason}`);
      break;
    }
  }

  return accumulatedSegments.join('');
}

/**
 * Stream Claude tokens in real time and extract stop_reason upon stream completion.
 */
export async function streamWithStopReasonCheck(
  client: Anthropic,
  prompt: string,
  maxTokens = 500,
  model = 'claude-3-5-sonnet-20241022'
): Promise<string> {
  const stream = await client.messages.create({
    model,
    max_tokens: maxTokens,
    stream: true,
    messages: [{ role: 'user', content: prompt }],
  });

  let fullText = '';

  for await (const event of stream) {
    if (event.type === 'content_block_delta' && event.delta.type === 'text_delta') {
      process.stdout.write(event.delta.text);
      fullText += event.delta.text;
    } else if (event.type === 'message_delta') {
      if (event.delta.stop_reason) {
        console.log(`\n[Stream Message Delta] Stop Reason: ${event.delta.stop_reason}`);
      }
    }
  }

  return fullText;
}

async function main(): Promise<void> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    console.error('ERROR: ANTHROPIC_API_KEY is not set in environment.');
    process.exit(1);
  }

  const client = new Anthropic({ apiKey });

  console.log('=== Demo 1: Triggering and Inspecting max_tokens Cutoff ===');
  const shortResp = await client.messages.create({
    model: 'claude-3-5-sonnet-20241022',
    max_tokens: 20,
    messages: [
      {
        role: 'user',
        content: 'Explain the Paxos consensus algorithm step by step.',
      },
    ],
  });
  inspectStopReason(shortResp);

  console.log('\n=== Demo 2: Triggering stop_sequence ===');
  const seqResp = await client.messages.create({
    model: 'claude-3-5-sonnet-20241022',
    max_tokens: 100,
    stop_sequences: ['--HALT--'],
    messages: [
      {
        role: 'user',
        content: 'Count numbers 1, 2, 3, 4, then insert --HALT--, then count 5, 6, 7.',
      },
    ],
  });
  inspectStopReason(seqResp);

  console.log('\n=== Demo 3: Automated Continuation Loop on max_tokens ===');
  const consolidated = await generateWithContinuation({
    client,
    prompt: 'Write an exhaustive 4-part guide detailing B-tree database indexing mechanics.',
    maxTokensPerCall: 80,
    maxRounds: 3,
  });

  console.log('\n--- Final Consolidated Output (Preview 200 chars) ---');
  console.log(consolidated.slice(0, 200) + '...');
}

if (require.main === module) {
  main().catch((err) => {
    console.error('Execution failed:', err);
    process.exit(1);
  });
}
